[PP] Add schedule visualizer

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #151249
* #151248
* #150359
* __->__ #150347

Added a new private file (`_schedule_visualizer.py`) with some helper methods that can be used to visualize the operations of a schedule and plot with matplotlib.

InterleavedZeroBubble(pp_group=4, microbatches=8):
![image](https://github.com/user-attachments/assets/610ba9a8-7d18-4a99-bcad-6f43e5b23c8c)




cc @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

H-Huang committed:
Add schedule visualizer

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] Add schedule visualizer"


Added a new private file (`_schedule_visualizer.py`) with some helper methods that can be used to visualize the operations of a schedule and plot with matplotlib.

InterleavedZeroBubble(pp_group=4, microbatches=8):
![image](https://github.com/user-attachments/assets/610ba9a8-7d18-4a99-bcad-6f43e5b23c8c)




cc awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/distributed/pipelining/_schedule_visualizer.py:
```diff
@@ -0,0 +1,176 @@
+"""
+
+from typing import Optional, Union
+from unittest import mock
```
what does this dependency do? Would this add a new dependency to pytorch lib?

H-Huang replied:
unittest is part of python's stdlib so it wont cause an issue!

kwen2501 commented on torch/distributed/pipelining/_schedule_visualizer.py:
```diff
@@ -0,0 +1,176 @@
+        plt.show()
+
+
+if __name__ == "__main__":
+    # Example usage:
+    ops = get_schedule_ops("InterleavedZeroBubble", 4, 8)
+    visualize_schedule(ops, "test.png")
```
Maybe we can write the example in comment format than under "__main__" ?

kwen2501 commented on torch/distributed/pipelining/_schedule_visualizer.py:
```diff
@@ -0,0 +1,176 @@
+from torch.distributed.pipelining.stage import PipelineStage
+
+
+def get_schedule_ops(
+    schedule: Union[str, _PipelineSchedule],
+    pp_degree: int,
+    num_microbatches: int,
+    num_stages_per_rank: Optional[int] = None,
+) -> list[list[Optional[_Action]]]:
```
Can you add a short comment to this function? thanks!

H-Huang replied:
will add, thanks

kwen2501 commented on torch/distributed/pipelining/_schedule_visualizer.py:
```diff
@@ -0,0 +1,176 @@
+}
+
+
+def visualize_schedule(
+    schedule: list[list[Optional[_Action]]], filename: Optional[str] = None
+) -> None:
```
Same here

H-Huang replied:
will do

kwen2501 reviewed (approved):
Looks great! Thank you!

H-Huang committed:
Update on "[PP] Add schedule visualizer"


Added a new private file (`_schedule_visualizer.py`) with some helper methods that can be used to visualize the operations of a schedule and plot with matplotlib.

InterleavedZeroBubble(pp_group=4, microbatches=8):
![image](https://github.com/user-attachments/assets/610ba9a8-7d18-4a99-bcad-6f43e5b23c8c)




cc awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge

H-Huang commented:
@pytorchbot merge -i
